src/jit/MemoryUtilInl.h

```diff
@@ -147,3 +147,143 @@ static void emitDataDrop(sljit_compiler* compiler, Instruction* instr)
     sljit_sw addr = GET_FUNC_ADDR(sljit_sw, dropData);
     sljit_emit_icall(compiler, SLJIT_CALL, SLJIT_ARGS2V(32, W), SLJIT_IMM, addr);
 }
+
+#if defined(ENABLE_EXTENDED_FEATURES)
+
+static void atomicNotifyCallback(ExecutionContext* context, sljit_s32 memOffset)
+{
+    Instance* instance = context->instance;
+    Memory** memories = reinterpret_cast<Memory**>(reinterpret_cast<uintptr_t>(instance) + Instance::alignedSize());
+    uint32_t result = 0;
+    auto count = context->tmp1[0];
+    memories[0]->atomicNotify(context->state, instance->module()->store(), memOffset, 0, count, &result);
+    context->tmp2[0] = result;
+}
+
+static void atomicWaitCallback(ExecutionContext* context, sljit_s32 memOffset, sljit_s32 size)
+{
+    Instance* instance = context->instance;
+    Memory** memories = reinterpret_cast<Memory**>(reinterpret_cast<uintptr_t>(instance) + Instance::alignedSize());
+    if (!memories[0]->isShared()) {
+        context->tmp1[0] = 1;
+        return;
+    }
+    uint32_t result = 0;
+    int64_t timeout = context->tmp2[0];
+    int64_t expect = context->tmp1[0];
+    if (size == 8) {
+        memories[0]->atomicWait(context->state, instance->module()->store(), memOffset, 0, expect, timeout, &result);
+    } else {
+        memories[0]->atomicWait(context->state, instance->module()->store(), memOffset, 0, (int32_t)expect, timeout, &result);
+    }
+    context->tmp2[0] = result;
+    context->tmp1[0] = 0;
+}
+
+static void emitAtomicSync(sljit_compiler* compiler, Instruction* instr)
```

Comment:
The code does not share any parts, it would be better to have 3 different functions.